scripts/Upgrade-PowerShell.ps1

```diff
@@ -167,62 +170,68 @@ Function Download-File($url, $path) {
-    Set-ItemProperty -Path $reg_winlogon_path -Name AutoAdminLogon -Value 0
-    Remove-ItemProperty -Path $reg_winlogon_path -Name DefaultUserName -ErrorAction SilentlyContinue
-    Remove-ItemProperty -Path $reg_winlogon_path -Name DefaultPassword -ErrorAction SilentlyContinue
+Function Set-AutoLogon {
+    param (
+        [switch] $Enable,
+        [switch] $Disable
+    )
+    $reg_winlogon_path = 'HKLM:\Software\Microsoft\Windows NT\CurrentVersion\Winlogon'
+
+    if ($Enable) {
+        Write-Log -Message 'setting auto logon registry properties'
+        Set-ItemProperty -Path $reg_winlogon_path -Name AutoAdminLogon -Value 1
+        Set-ItemProperty -Path $reg_winlogon_path -Name DefaultUserName -Value $username
+        Set-ItemProperty -Path $reg_winlogon_path -Name DefaultPassword -Value $password
+        if ($domain) {
+            Set-ItemProperty -Path $reg_winlogon_path -Name DefaultDomain -Value $domain
+        }
+    } elseif ($Disable) {
+        Write-Log -Message 'clearing auto logon registry properties'
+        Set-ItemProperty -Path $reg_winlogon_path -Name AutoAdminLogon -Value 0
+        Remove-ItemProperty -Path $reg_winlogon_path -Name DefaultUserName -ErrorAction SilentlyContinue
+        Remove-ItemProperty -Path $reg_winlogon_path -Name DefaultPassword -ErrorAction SilentlyContinue
+        Remove-ItemProperty -Path $reg_winlogon_path -Name DefaultDomain -ErrorAction SilentlyContinue
+    } else {
+        $error_msg = 'ambiguous calling of Set-Autolon: expecting a parameter -Enable or -Disable, none has been provided'
+        Write-Log -Message $error_msg -Level 'ERROR'
+        throw $error_msg
+    }
 }
 
 Function Download-Wmf5Server2008($architecture) {
     if ($architecture -eq "x64") {
-        $zip_url = "http://download.microsoft.com/download/6/F/5/6F5FF66C-6775-42B0-86C4-47D41F2DA187/Win7AndW2K8R2-KB3191566-x64.zip"
-        $file = "$tmp_dir\Win7AndW2K8R2-KB3191566-x64.msu"
+        $zip_url = [uri]"http://download.microsoft.com/download/6/F/5/6F5FF66C-6775-42B0-86C4-47D41F2DA187/Win7AndW2K8R2-KB3191566-x64.zip"
     } else {
-        $zip_url = "http://download.microsoft.com/download/6/F/5/6F5FF66C-6775-42B0-86C4-47D41F2DA187/Win7-KB3191566-x86.zip"
-        $file = "$tmp_dir\Win7-KB3191566-x86.msu"
+        $zip_url = [uri]"http://download.microsoft.com/download/6/F/5/6F5FF66C-6775-42B0-86C4-47D41F2DA187/Win7-KB3191566-x86.zip"
     }
+    $filename = $zip_url.Segments[-1]
+    $file = Join-Path -Path "$tmp_dir" -ChildPath $filename.Replace('.zip', '.msu')
     if (Test-Path -Path $file) {
         return $file
     }
 
-    $filename = $zip_url.Split("/")[-1]
-    $zip_file = "$tmp_dir\$filename"
+    $zip_file = Join-Path -Path "$tmp_dir" -ChildPath "$filename"
     Download-File -url $zip_url -path $zip_file
 
     Write-Log -message "extracting '$zip_file' to '$tmp_dir'"
-    try {
-        Add-Type -AssemblyName System.IO.Compression.FileSystem > $null
```

Comment:
Please keep this in here, Server 2008/R2 probably won't include .NET 4.5 which is when `System.IO.Compression.ZipFile` was added.

Reply:
You are correct that .NET 4.5 isn't available in clean Server 2008 R2, and you definitely found a bug in my code, but my intentions was to get rid of the `CopyHere()` method from `Shell.Application` object as it is unreliable and unpredictable. 
We could achieve that by loading assembly in a slightly different way and also telling PowerShell to use latest installed NET Framework - it will work because first step in upgrading PS to version 3.0 and highter is to install .NET 4.5.2. You can find more information in that SO answer: https://stackoverflow.com/a/37815418/10504393

So, after .NET installation we could add DWORD that forces usage of the latest CRL, respawn script and load assembly like this:
```powershell
[Reflection.Assembly]::LoadWithPartialName("System.IO.Compression.Filesystem")
```

Also, this could allow us to get rid of `Download-Wmf5Server2008` and rewrite script in more consistent way, that all steps logic will be determined in `foreach ($action in $actions)` loop. 

What do you think?